spectral_cube/spectral_cube.py

```diff
@@ -2404,7 +2375,257 @@ def reproject(self, header, order='bilinear'):
+
+                self._try_load_beam(header)
+
+                if hasattr(self, 'beam') or hasattr(self, 'beams'):
+                    warnings.warn("Units were Jy/beam.  The 'beam' is now "
+                                  "stored in the .beam attribute, and the "
+                                  "units are set to Jy")
+                else:
+                    warnings.warn("Could not parse Jy/beam unit.  Either "
+                                  "you should install the radio_beam "
+                                  "package or manually replace the units."
+                                  " For now, the units are being interpreted "
+                                  "as Jy.")
+
+        if beam is not None:
+            self.beam = beam
+
+        if 'beam' in self._meta:
+            self.pixels_per_beam = (self.beam.sr /
+                                    (astropy.wcs.utils.proj_plane_pixel_area(self.wcs) *
+                                     u.deg**2)).to(u.dimensionless_unscaled).value
+
+
+    @property
+    def beam(self):
+        return self._meta['beam']
+
+    @beam.setter
+    def beam(self, value):
+        self._meta['beam'] = value
+
+    def spectral_smooth(self, kernel,
+                        #numcores=None,
+                        convolve=convolution.convolve,
+                        **kwargs):
+        """
+        Smooth the cube along the spectral dimension
+
+        Parameters
+        ----------
+        kernel : `~astropy.convolution.Kernel1D`
+            A 1D kernel from astropy
+        numcores : int
+            Number of cores to use in parallel-processing.
+        convolve : function
+            The astropy convolution function to use, either
+            `astropy.convolution.convolve` or
+            `astropy.convolution.convolve_fft`
+        kwargs : dict
+            Passed to the convolve function
+        """
+
+        shape = self.shape
+
+        # "cubelist" is a generator
+        # the boolean check will skip smoothing for bad spectra
+        # TODO: should spatial good/bad be cached?
+        cubelist = ((self.filled_data[:,jj,ii],
+                     self.mask.include(view=(slice(None), jj, ii)))
+                    for jj in xrange(self.shape[1])
+                    for ii in xrange(self.shape[2]))
+
+        pb = ProgressBar(shape[1]*shape[2])
+
+        def _gsmooth_spectrum(args):
+            """
+            Helper function to smooth a spectrum
+            """
+            (spec, includemask),kernel,kwargs = args
+            pb.update()
+
+            if any(includemask):
+                return convolve(spec, kernel, normalize_kernel=True, **kwargs)
+            else:
+                return spec
+
+        # could be numcores, except _gsmooth_spectrum is unpicklable
+        with cube_utils._map_context(1) as map:
+            smoothcube_ = np.array([x for x in
+                                    map(_gsmooth_spectrum, zip(cubelist,
```

Comment:
`itertools.izip` instead? I think zip evaluates the whole generator at once (or is that just python2?).